### DISCUSSION TITLE
Trying to emulate hitbox controller rules

### DISCUSSION THREAD

**Discussion by zehthales** [sezanzeb/input-remapper#465] [Q&A]
I'm playing Street Fighter V on steam with my keyboard. The steam key mapping for Keyboard do something like the key left pressed together with key right have the output of always following the opponent position. 

At hitbox controller is simpler, right cancels left and up cancel down. Basically right+left=down+up=stand. 

I tried to something like this with input-remaper, but no success by now. 

Note: the remapping works inside the game, I just don't know how to implement it

**Comment by jonasBoss** [comment]
Did I get this correct: if you press left+right or up+down you want the output to be no key pressed?

This is a bit difficult on the current release, but if you switch to [beta](https://github.com/sezanzeb/input-remapper/tree/beta) you can do this:
- `left+right` -> `disable` (literally just write "disable" in the output field)
- `up+down` -> `disable`

On the current release you will probably need 6 mappings:

- `left` -> `KEY_LEFT`
- `right` -> `KEY_RIGHT` 
- `left+right` -> `event(EV_KEY, KEY_LEFT, 0).event(EV_KEY, KEY_RIGHT, 0)` this will release both the left an right key
- the same for up and down

**Reply by zehthales** [reply]
Oh! Beta looks very nice! I'll try it.

I playing on steam and mapped left to KEY_A and right to KEY_D and i tried something like this:

a+d  mapped to u (not used by the game) 
d+a mapped to u also

The problem was if i'm holding "a" and  press "d", after releasing "a" while holding down "d" the character stuck. Same thing to the opposite situation.

**Reply by zehthales** [reply]
Also, beta not available on this github link. How can I install it?

**Reply by jonasBoss** [reply]
> The problem was if i'm holding "a" and press "d", after releasing "a" while holding down "d" the character stuck.

That is a whole different problem, which won't be easy to solve. Maybe someone can come up with a macro utilizing the macro variables.
The issue is, that in order to stop the character both `a` and `d` keys are released. So we need to press `a` or `d` again as soon as the other key is actually released. This requires us to know which key was released. 
I am afraid this is actually not possible because a macro does not know what keys are pressed.
But I can see how useful this would be, maybe we can change the macro system to allow this.

> How can I install it?

That depends on your system if you have a Arch based distro there is a AUR for this.
On any other system you will need to follow the manual install process:
```bash
sudo pip install evdev -U  # If newest version not in distros repo
sudo pip uninstall key-mapper  # In case the old package is still installed
sudo pip install --no-binary :all: git+https://github.com/sezanzeb/input-remapper.git@beta
sudo systemctl enable input-remapper
sudo systemctl restart input-remapper
```
make sure to uninstall input-remapper first.

**Reply by zehthales** [reply]
That's sad. Probably with one of the  if_command is possible to do this. But is really hard. At a+d = w, I could add the info "key_up(a) ends macro" or something like that.

About the beta version, I was able to install, but the GUI app doesn't record anything and the  `sudo systemctl restart input-remapper`  I get an error.

```
input-remapper.service - Service to inject keycodes without the GUI application
     Loaded: loaded (/lib/systemd/system/input-remapper.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Thu 2022-09-01 17:17:46 -03; 2min 18s ago
    Process: 47177 ExecStart=/usr/bin/input-remapper-service (code=exited, status=9)
   Main PID: 47177 (code=exited, status=9)
        CPU: 241ms

Sep 01 17:17:46 pop-os systemd[1]: Starting Service to inject keycodes without the GUI application...
Sep 01 17:17:46 pop-os input-remapper-service[47177]: input-remapper-service 1.6.0-beta 5f98c0941dc7fdead51c8047bf682c1e06f7>
Sep 01 17:17:46 pop-os input-remapper-service[47177]: python-evdev 1.6.0
Sep 01 17:17:46 pop-os input-remapper-service[47177]: ERROR: Is the service already running? (name already exists on the bus)
Sep 01 17:17:46 pop-os input-remapper-service[47177]: Stopping all injections
Sep 01 17:17:46 pop-os systemd[1]: input-remapper.service: Main process exited, code=exited, status=9/n/a
Sep 01 17:17:46 pop-os systemd[1]: input-remapper.service: Failed with result 'exit-code'.
Sep 01 17:17:46 pop-os systemd[1]: Failed to start Service to inject keycodes without the GUI application.
```

**Reply by jonasBoss** [reply]
> `Sep 01 17:17:46 pop-os input-remapper-service[47177]: ERROR: Is the service already running? (name already exists on the bus)`

It seems like there is still some old input-remapper running. Try `$sudo pkill input-remapper-` and then start it again.

**Reply by zehthales** [reply]
new error 

Failed to enable unit: Unit file /etc/systemd/system/input-remapper-daemon.service is masked.

**Reply by zehthales** [reply]
Failed to restart input-remapper.service: Unit input-remapper-daemon.service is masked.

**Reply by jonasBoss** [reply]
The symlinks in `/etc/systemd/system/input-remapper-daemon.service` are probably broken due to the manual installation. 
You can unmask the service with `$sudo systemctl unmask input-remapper.service`. input-remapper will no longer start automatically on system startup, for that you need to enable it again: `$sudo systemctl enable input-remapper.service`

**Reply by zehthales** [reply]
The unmask command had no output. After this, trying to enable says that still masked, but the beta version is fully working now. I even tried you tip on remapping keyboard into gamepad analog or axis

Thank you very much!

**Reply by jonasBoss** [reply]
I think on beta this is as close as it gets:
```json
{
    "1,17,1": {
        "target_uinput": "keyboard",
        "output_symbol": "set(y_state, 1)\n.key_down(KEY_UP)\n.hold(\n\tif_eq(\n\t\t$y_state, 0,\n\t\tset(y_state, 1).key_down(KEY_UP),\n\t)\n)\n.if_eq(\n\t$y_state, 0,\n\tset(y_state, -1).key_down(KEY_DOWN),\n\tkey_up(KEY_UP)\n)"
    },
    "1,31,1": {
        "target_uinput": "keyboard",
        "output_symbol": "set(y_state, -1)\n.key_down(KEY_DOWN)\n.hold(\n\tif_eq(\n\t\t$y_state, 0,\n\t\tset(y_state, -1).key_down(KEY_DOWN),\n\t)\n)\n.if_eq(\n\t$y_state, 0,\n\tset(y_state, 1).key_down(KEY_UP),\n\tkey_up(KEY_DOWN)\n)"
    },
    "1,17,1+1,31,1": {
        "target_uinput": "keyboard",
        "output_symbol": "key_up(KEY_UP).key_up(KEY_DOWN)\n.hold()\n.set(y_state, 0)"
    }
}

```
This will map the "w" and "s" keys to the up and down arrow keys. There is one issue: 
pressing "w" then "s" and then releasing "w" will result in a stuck down arrow until "s" is pressed and released again. 

same thing but with the gamepad axis instead:
```json
{
    "1,30,1": {
        "target_uinput": "gamepad",
        "output_symbol": "set(x_state, 1)\n.event(3,0,32768)\n.hold(\n\tif_eq(\n\t\t$x_state, 0,\n\t\tset(x_state, 1).event(3,0,32768),\n\t)\n)\n.if_eq(\n\t$x_state, 0,\n\tset(x_state, -1).event(3,0,-32768),\n\tevent(3,0,0)\n)"
    },
    "1,32,1": {
        "target_uinput": "gamepad",
        "output_symbol": "set(x_state, -1)\n.event(3,0,-32768)\n.hold(\n\tif_eq(\n\t\t$x_state, 0,\n\t\tset(x_state, -1).event(3,0,-32768),\n\t)\n)\n.if_eq(\n\t$x_state, 0,\n\tset(x_state, 1).event(3,0,32768),\n\tevent(3,0,0)\n)\n"
    },
    "1,30,1+1,32,1": {
        "target_uinput": "gamepad",
        "output_symbol": "event(3,0,0)\n.hold()\n.set(x_state, 0)"
    }
}
```

**Reply by zehthales** [reply]
This is very close! Solve most of the buffers problem related with steam mapping. If we could add a releasing information to input key to avoid that problem...

**Reply by zehthales** [reply]
What happens if I write "1,30,1+1,32,1+1,30,0" ?

**Reply by jonasBoss** [reply]
you will get an error or the mapping will just not work.

I am very certain that you can not write a macro that solves this entirely. Because the macro is missing information about the keys which are pressed and released. 
The solution above is trying to infer information by observing when which mapping is triggered and released. 
If we press "w" then "s" and then release "w" no more macro is running and we end up with a stuck key.

**Comment by zehthales** [comment]
too bad! But your solution was very nice! Thank you for the help
